Source/JavaScriptCore/runtime/WaiterListManager.h

```diff
@@ -202,13 +202,41 @@ class WaiterList : public ThreadSafeRefCounted<WaiterList> {
     SentinelLinkedList<Waiter, BasicRawSentinelNode<Waiter>> m_waiters;
 };
 
+class WaiterListManager;
+
+struct LockedWaiterList {
+    LockedWaiterList(WaiterListManager& manager, Ref<WaiterList>&& list, void* ptr)
+        : m_manager(manager)
+        , m_list(list)
+        , m_locker(list->lock)
+        , m_address(ptr)
+    { }
+
+    LockedWaiterList(const LockedWaiterList&) = delete;
+    LockedWaiterList(LockedWaiterList&&) = delete;
+
+    LockedWaiterList& operator=(const LockedWaiterList&) = delete;
+    LockedWaiterList& operator=(LockedWaiterList&&) = delete;
+
+    bool waitSync(VM&, Seconds timeout);
+    void waitAsync(Ref<Waiter>, Seconds timeout);
+
+private:
+    WaiterListManager& m_manager;
+    Ref<WaiterList> m_list;
+    Locker<WTF::Lock> m_locker;
```

Comment:
Locker should not be in a member of some objects since this is really dangerous idiom for locking.